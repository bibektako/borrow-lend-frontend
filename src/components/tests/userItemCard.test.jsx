// src/components/__tests__/UserItemCard.test.jsx
import React from 'react';
import { describe, it, expect, vi } from 'vitest';
import { render, screen, fireEvent } from '@testing-library/react';
import UserItemCard from '../user/User_Items_card';
import { getBackendImageUrl } from "../../../utils/backend-image"

vi.mock('../../../utils/backend-image', () => ({
  getBackendImageUrl: vi.fn(() => 'https://mocked.image.url/test.jpg'),
}));

const mockItem = {
  _id: 'item123',
  name: 'Test Item',
  borrowingPrice: 250,
  imageUrls: ['test.jpg'],
  category: { name: 'Tools' },
};

describe('UserItemCard', () => {
  const mockOnEdit = vi.fn();
  const mockOnDelete = vi.fn();

  beforeEach(() => {
    vi.clearAllMocks();
  });

  it('renders item details correctly', () => {
    render(<UserItemCard item={mockItem} onEdit={mockOnEdit} onDelete={mockOnDelete} />);

    expect(screen.getByText('Tools')).toBeInTheDocument();
    expect(screen.getByText('Test Item')).toBeInTheDocument();
    expect(screen.getByText('Rs 250')).toBeInTheDocument();
    expect(screen.getByRole('img')).toHaveAttribute('src', 'https://mocked.image.url/test.jpg');
  });

  it('calls onEdit when Edit button is clicked', () => {
    render(<UserItemCard item={mockItem} onEdit={mockOnEdit} onDelete={mockOnDelete} />);

    fireEvent.click(screen.getByText('Edit'));
    expect(mockOnEdit).toHaveBeenCalledWith(mockItem);
  });

  it('calls onDelete when Delete button is clicked', () => {
    render(<UserItemCard item={mockItem} onEdit={mockOnEdit} onDelete={mockOnDelete} />);

    fireEvent.click(screen.getByText('Delete'));
    expect(mockOnDelete).toHaveBeenCalledWith(mockItem._id);
  });

  it('displays "Uncategorized" if no category exists', () => {
    const uncategorizedItem = { ...mockItem, category: null };
    render(<UserItemCard item={uncategorizedItem} onEdit={mockOnEdit} onDelete={mockOnDelete} />);
    expect(screen.getByText('Uncategorized')).toBeInTheDocument();
  });

  it('uses fallback image if imageUrls is empty', () => {
    const noImageItem = { ...mockItem, imageUrls: [] };
    render(<UserItemCard item={noImageItem} onEdit={mockOnEdit} onDelete={mockOnDelete} />);
    expect(screen.getByRole('img')).toHaveAttribute(
      'src',
      'https://placehold.co/600x400/E2E8F0/334155?text=No+Image'
    );
  });

  it('matches snapshot', () => {
    const { asFragment } = render(<UserItemCard item={mockItem} onEdit={mockOnEdit} onDelete={mockOnDelete} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
