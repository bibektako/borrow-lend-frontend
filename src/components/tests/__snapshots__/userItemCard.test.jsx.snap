// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UserItemCard > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="bg-white rounded-lg shadow-md overflow-hidden flex flex-col border border-gray-200"
  >
    <div
      class="w-full h-48 bg-gray-200 flex items-center justify-center overflow-hidden"
    >
      <img
        alt="Test Item"
        class="w-full h-full object-cover"
        src="https://mocked.image.url/test.jpg"
      />
    </div>
    <div
      class="p-4 flex flex-col flex-grow"
    >
      <p
        class="text-xs text-blue-600 font-semibold"
      >
        Tools
      </p>
      <h3
        class="text-md font-bold text-gray-800 mt-1 truncate"
      >
        Test Item
      </h3>
      <p
        class="text-lg font-black text-gray-900 mt-2"
      >
        Rs 250
      </p>
      <div
        class="mt-auto pt-4 flex gap-3"
      >
        <button
          class="flex-1 bg-gray-100 text-gray-800 font-semibold py-2 rounded-lg hover:bg-gray-200 transition-colors"
        >
          Edit
        </button>
        <button
          class="flex-1 bg-red-500 text-white font-semibold py-2 rounded-lg hover:bg-red-600 transition-colors"
        >
          Delete
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
